respect global strip config

malt3 opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, please read CONTRIBUTING.md and sign the CLA
   first. We cannot review code without a signed CLA.
2. Please file an issue *first*. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
-->

**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

The docs currently state:

> Strips symbols from compiled packages and linked binaries (using the ``-w`` flag). May also be set with the ``--strip`` command line option, which affects C/C++ targets, too.

This is not true. This option cannot be set at all (as far as I can tell). It is currently not exposed as a config flag and the global bazel option for stripping is not considered either.
This PR removes the `@io_bazel_rules_go//go/config:strip` config setting and always uses the global bazel setting `--strip` and `--compilation_mode` to deduce if stripping should be enabled.
As a little extension, I also changed the linker flags to also include `-s` if stripping is enabled.

**Which issues(s) does this PR fix?**

Fixes #3325

**Other notes for review**

This is one possible resolution. See also #3526 for an alternative solution that exposes this setting as a flag.

fmeum reviewed (approved):
This looks good to me, including the addition of the `-s` flag. 

Given the very terse descriptions of the `-s` and `-w` flags in the `go tool link` documentation, could you add a test that verifies that stack traces in panics still have all the expected information? 

Ideally that test would also check that stripping is enabled when it should be, which would probably have caught the typo below.

malt3 commented on go/private/BUILD.bazel:
```diff
@@ -30,6 +30,21 @@ config_setting(
+config_setting(
+    name = "is_strip_sometimes_fastbuild",
+    values = {
+        "strip": "always",
```
Good catch

malt3 commented:
>  could you add a test that verifies that stack traces in panics still have all the expected information

Yes! I'll try to get to it very soon.

malt3 commented:
I implemented tests for ELF, MACH-O and PE. The checks for debug symbols in the resulting binaries are not very good and could certainly be improved if desired. I'm not sure where to go with this or what to test for in particular.

I also added tests to trigger a panic, recover from it and compare the stack trace against a known good stack trace.

fmeum commented on tests/core/strip/strip_test.go:
```diff
@@ -0,0 +1,290 @@
+	}{
+		{
+			desc:      "run_auto",
+			cmd:       "run",
```
This seems to be the only value, could this be inlined?

fmeum reviewed (approved):
This is great test coverage. That will make it much easier to maintain this feature.

I see that the windows run is failing, but everything else looks good.

linzhp commented:
Thanks for letting me know. We can pull this into Uber's Go repo and test it

fmeum commented:
We should definitely mention this in the release notes. Since we are preserving debug information and `-c opt` builds with default settings aren't affected, I would have said that's sufficient, but would like to hear what @linzhp and @achew22 think.

linzhp commented:
Should we also pass `-dwarf=false` to the Go compiler when `go.mode.strip` is True?

malt3 commented:
I'm not sure I can make a statement on the behavior with regards to the bazel cache when setting `-dwarf=false`.
Or to rephrase:
Could this lead to missing dwarf info if I first build a stripped binary and then rebuild with debugging info? Maybe someone more experienced with the internals can comment.

This discussion has some stats on the performance impact of dwarf generation: https://github.com/golang/go/issues/38777
Not a huge difference, but not nothing either so I'd be happy to add the flag.

linzhp reviewed (approved):
All tests passed in Uber, after some fixes removing the usage of the old strip flag, and using transitions to dbg mode when debug symbols are needed.
